client-java/src/main/java/uk/gov/gchq/palisade/client/ClientManager.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2018-2021 Crown Copyright
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package uk.gov.gchq.palisade.client;
+
+import uk.gov.gchq.palisade.client.internal.dft.DefaultClient;
+
+import java.util.Map;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+/**
+ * The basic service for managing Palisade clients
+ * <ul>
+ * <li>pal://host:port - standard client (which using http for PS and FRS</li>
+ * <li>pal:dft://host:port - standard client (which using http for PS and
+ * FRS</li>
+ * <li>pal:alt://host:port - alternative client</li>
+ * </ul>
+ *
+ * @since 0.5.0
+ */
+public final class ClientManager {
+
+    private static final CopyOnWriteArrayList<Client> REGISTERED_CLIENTS = new CopyOnWriteArrayList<>();
```

Comment:
The list is mainly only read so use the `CopyOnWriteArrayList` is preferable in this use case as no `synchronized` is needed for reading. Indeed the `CopyOnWriteArrayList` does not perform any locking for reading data.

Reply:
The emphasis was really on cleaner, how often are you doing a get for a new registered client?
If performance is a concern, `ReadWriteReentrantLock` is what you'd want rather than this.

Reply:
`CopyOnWriteArrayList` uses a `ReentrantLock` internally when you do a write only. When performing reads, no locking is performed. If I use a `HashSet`, I would have to put synchronised blocks everywhere the set is read. As this is usually initialised once and then read many, this is the correct with only a single synchronized block.

Reply:
I've added some documentation to make it clearer the intention of this code and why it's like it is.